File: ChainOfResponsibilityPattern/Chain of Responsibility Pattern.cpp
#include <iostream>
#include <string>

using namespace std;

class Data {
public:
    string info;
    string format;
    int sizeInMB;
    bool containsPersonalInformation;

    Data(
        string info,
        string format,
        int sizeInMB,
        bool containsPersonalInformation
    ) {
        this->info = info;
        this->format = format;
        this->sizeInMB = sizeInMB;
        this->containsPersonalInformation = containsPersonalInformation;
    }
};

class IHandler {
public:
    virtual ~IHandler() = default;

    virtual IHandler* SetNext(IHandler* nextHandler) = 0;

    virtual bool Handle(const Data& data) = 0;
};

class BaseDataHandler : public IHandler {
private:
    IHandler* nextHandler = nullptr;

public:
    IHandler* SetNext(IHandler* nextHandler) override {
        this->nextHandler = nextHandler;
        return nextHandler;
    }

    bool Handle(const Data& data) override {
        if (nextHandler != nullptr) {
            return nextHandler->Handle(data);
        }

        return true;
    }
};

class ValidationChecks : public BaseDataHandler {
public:
    bool Handle(const Data& data) override {
        cout << "Running Validation Check..." << endl;

        if (data.info.empty()) {
            cout << "Validation failed: data is empty." << endl;
            return false;
        }

        cout << "Validation passed." << endl;
        return BaseDataHandler::Handle(data);
    }
};

class FormattingChecks : public BaseDataHandler {
public:
    bool Handle(const Data& data) override {
        cout << "Running Formatting Check..." << endl;

        if (data.format != "CSV" && data.format != "JSON") {
            cout << "Formatting failed: unsupported format." << endl;
            return false;
        }

        cout << "Formatting passed." << endl;
        return BaseDataHandler::Handle(data);
    }
};

class DataSizeCheck : public BaseDataHandler {
private:
    int maxSizeInMB;

public:
    DataSizeCheck(int maxSizeInMB) {
        this->maxSizeInMB = maxSizeInMB;
    }

    bool Handle(const Data& data) override {
        cout << "Running Data Size Check..." << endl;

        if (data.sizeInMB > maxSizeInMB) {
            cout << "Data size failed: data is too large." << endl;
            return false;
        }

        cout << "Data size passed." << endl;
        return BaseDataHandler::Handle(data);
    }
};

class PersonalInformationChecks : public BaseDataHandler {
public:
    bool Handle(const Data& data) override {
        cout << "Running Personal Information Check..." << endl;

        if (data.containsPersonalInformation) {
            cout << "Personal information check failed: sensitive data found." << endl;
            return false;
        }

        cout << "Personal information check passed." << endl;
        return BaseDataHandler::Handle(data);
    }
};

class BatchJob {
private:
    IHandler* pipeline;

public:
    BatchJob(IHandler* pipeline) {
        this->pipeline = pipeline;
    }

    void Process(const Data& data) {
        cout << "Starting batch job pipeline..." << endl;
        cout << "-----------------------------" << endl;

        bool passed = pipeline->Handle(data);

        cout << "-----------------------------" << endl;

        if (passed) {
            cout << "All checks passed. Processing data..." << endl;
        } else {
            cout << "Pipeline failed. Data processing stopped." << endl;
        }

        cout << endl;
    }
};

int main() {
    ValidationChecks validation;
    FormattingChecks formatting;
    DataSizeCheck dataSize(100);
    PersonalInformationChecks personalInfo;

    validation
        .SetNext(&formatting)
        ->SetNext(&dataSize)
        ->SetNext(&personalInfo);

    BatchJob batchJob(&validation);

    Data validData(
        "Big data content",
        "CSV",
        50,
        false
    );

    Data invalidFormatData(
        "Some content",
        "TXT",
        20,
        false
    );

    Data largeData(
        "Huge data content",
        "JSON",
        150,
        false
    );

    Data personalData(
        "Customer private data",
        "CSV",
        30,
        true
    );

    cout << "Test Case 1: Valid Data" << endl;
    batchJob.Process(validData);

    cout << "Test Case 2: Invalid Format Data" << endl;
    batchJob.Process(invalidFormatData);

    cout << "Test Case 3: Large Data" << endl;
    batchJob.Process(largeData);

    cout << "Test Case 4: Personal Information Data" << endl;
    batchJob.Process(personalData);

    return 0;
}

File: ChainOfResponsibilityPattern/Chain of Responsibility Pattern.cs
using System;

public class Data
{
    public string Info { get; set; }
    public string Format { get; set; }
    public int SizeInMB { get; set; }
    public bool ContainsPersonalInformation { get; set; }

    public Data(
        string info,
        string format,
        int sizeInMB,
        bool containsPersonalInformation
    )
    {
        Info = info;
        Format = format;
        SizeInMB = sizeInMB;
        ContainsPersonalInformation = containsPersonalInformation;
    }
}

public interface IHandler
{
    IHandler SetNext(IHandler nextHandler);

    bool Handle(Data data);
}

public abstract class BaseDataHandler : IHandler
{
    private IHandler? _nextHandler;

    public IHandler SetNext(IHandler nextHandler)
    {
        _nextHandler = nextHandler;
        return nextHandler;
    }

    public virtual bool Handle(Data data)
    {
        if (_nextHandler != null)
        {
            return _nextHandler.Handle(data);
        }

        return true;
    }
}

public class ValidationChecks : BaseDataHandler
{
    public override bool Handle(Data data)
    {
        Console.WriteLine("Running Validation Check...");

        if (string.IsNullOrWhiteSpace(data.Info))
        {
            Console.WriteLine("Validation failed: data is empty.");
            return false;
        }

        Console.WriteLine("Validation passed.");
        return base.Handle(data);
    }
}

public class FormattingChecks : BaseDataHandler
{
    public override bool Handle(Data data)
    {
        Console.WriteLine("Running Formatting Check...");

        if (data.Format != "CSV" && data.Format != "JSON")
        {
            Console.WriteLine("Formatting failed: unsupported format.");
            return false;
        }

        Console.WriteLine("Formatting passed.");
        return base.Handle(data);
    }
}

public class DataSizeCheck : BaseDataHandler
{
    private readonly int _maxSizeInMB;

    public DataSizeCheck(int maxSizeInMB)
    {
        _maxSizeInMB = maxSizeInMB;
    }

    public override bool Handle(Data data)
    {
        Console.WriteLine("Running Data Size Check...");

        if (data.SizeInMB > _maxSizeInMB)
        {
            Console.WriteLine("Data size failed: data is too large.");
            return false;
        }

        Console.WriteLine("Data size passed.");
        return base.Handle(data);
    }
}

public class PersonalInformationChecks : BaseDataHandler
{
    public override bool Handle(Data data)
    {
        Console.WriteLine("Running Personal Information Check...");

        if (data.ContainsPersonalInformation)
        {
            Console.WriteLine("Personal information check failed: sensitive data found.");
            return false;
        }

        Console.WriteLine("Personal information check passed.");
        return base.Handle(data);
    }
}

public class BatchJob
{
    private readonly IHandler _pipeline;

    public BatchJob(IHandler pipeline)
    {
        _pipeline = pipeline;
    }

    public void Process(Data data)
    {
        Console.WriteLine("Starting batch job pipeline...");
        Console.WriteLine("-----------------------------");

        bool passed = _pipeline.Handle(data);

        Console.WriteLine("-----------------------------");

        if (passed)
        {
            Console.WriteLine("All checks passed. Processing data...");
        }
        else
        {
            Console.WriteLine("Pipeline failed. Data processing stopped.");
        }

        Console.WriteLine();
    }
}

public class Program
{
    public static void Main()
    {
        IHandler validation = new ValidationChecks();
        IHandler formatting = new FormattingChecks();
        IHandler dataSize = new DataSizeCheck(100);
        IHandler personalInfo = new PersonalInformationChecks();

        validation
            .SetNext(formatting)
            .SetNext(dataSize)
            .SetNext(personalInfo);

        BatchJob batchJob = new BatchJob(validation);

        Data validData = new Data(
            "Big data content",
            "CSV",
            50,
            false
        );

        Data invalidFormatData = new Data(
            "Some content",
            "TXT",
            20,
            false
        );

        Data largeData = new Data(
            "Huge data content",
            "JSON",
            150,
            false
        );

        Data personalData = new Data(
            "Customer private data",
            "CSV",
            30,
            true
        );

        Console.WriteLine("Test Case 1: Valid Data");
        batchJob.Process(validData);

        Console.WriteLine("Test Case 2: Invalid Format Data");
        batchJob.Process(invalidFormatData);

        Console.WriteLine("Test Case 3: Large Data");
        batchJob.Process(largeData);

        Console.WriteLine("Test Case 4: Personal Information Data");
        batchJob.Process(personalData);
    }
}